engine/src/core/error/expected.h

```diff
@@ -0,0 +1,108 @@
+//
+// Created by grzegorz on 18.09.19.
+//
+
+#pragma once
+
+#include <utility>
+#include "unexpected.h"
+#include "bald_assert.h"
+
+
+namespace Bald {
+    template<typename E, typename U>
+    class expected {
+        union {
+            E valid;
+            U invalid;
+        };
+        bool _isOk = true;
+    public:
+        constexpr expected() { new(&valid) E(); }
+
+        constexpr expected(const E& rhs) { new(&valid) E(rhs); }
+
+        constexpr expected(const unexpected<U>& rhs) :
+                _isOk(false) { new(&invalid) U(rhs.value); }
+
+        constexpr expected(E&& rhs) { new(&valid) E(std::forward<E>(rhs)); }
+
+        constexpr expected(unexpected<U>&& rhs) :
+                _isOk(false) { new(&invalid) U(std::forward<U>(rhs.value)); }
+
+        constexpr expected(const expected<E, U>& rhs) :
+                _isOk(rhs._isOk) {
+            if (rhs._isOk) new(&valid) E(rhs.valid);
+            else new(&invalid) U(rhs.invalid);
+        }
+
+        constexpr expected(expected<E, U>&& rhs) noexcept :
+                _isOk(rhs._isOk) {
+            if (rhs._isOk) new(&valid) E(std::move(rhs.valid));
+            else new(&invalid) U(std::move(rhs.invalid));
+        }
```

Comment:
Make all single argument constructors explicit. Also, check for noexcept correctness. I believe placement new can throw an exception, so take one more look at line 39.